File: HashTables/productExceptSelf.c
#include <stdio.h>
#include <stdlib.h>

/*
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.
*/

int *productExceptSelf(int *nums, int numsSize, int *returnSize)
{
    // Définir la taille du tableau résultat
    *returnSize = numsSize;
    
    int *res = (int *)malloc(sizeof(int) * numsSize);
    if (!res)
        return NULL;
        
    int *pref = (int *)malloc(sizeof(int) * numsSize);
    int *suf = (int *)malloc(sizeof(int) * numsSize);
    
    if (!pref || !suf) {
        free(res);
        free(pref); // Libération sécurisée (fonctionne même si pref est NULL)
        free(suf);
        return NULL;
    }

    // Initialisation
    for (int i = 0; i < numsSize; i++)
    {
        res[i] = pref[i] = suf[i] = 1;
    }

    // Calculer les préfixes
    for (int i = 1; i < numsSize; i++)
    {
        pref[i] = nums[i-1] * pref[i-1];
    }

    // Calculer les suffixes (correction pour inclure i=0)
    for(int i = numsSize - 2; i >= 0; i--)
    {
        suf[i] = suf[i+1] * nums[i+1];
    }

    // Calculer le résultat
    for(int i = 0; i < numsSize; i++)
    {
        res[i] = suf[i] * pref[i];
    }

    // Libérer la mémoire des tableaux temporaires
    free(pref);
    free(suf);

    return res;
}

File: HashTables/productExceptSelf.py
# Brute Force\
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n

        for i in range(n):
            prod = 1
            for j in range(n):
                if i == j:
                    continue    
                prod *= nums[j]
            
            res[i] = prod
        return res
    
# Division
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prod, zero_cnt = 1, 0
        for num in nums:
            if num:
                prod *= num
            else:
                zero_cnt +=  1
        if zero_cnt > 1: return [0] * len(nums)

        res = [0] * len(nums)
        for i, c in enumerate(nums):
            if zero_cnt: res[i] = 0 if c else prod
            else: res[i] = prod // c
        return res
    
    
#  Prefix & Suffix
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0] * n
        pref = [0] * n
        suff = [0] * n

        pref[0] = suff[n - 1] = 1
        for i in range(1, n):
            pref[i] = nums[i - 1] * pref[i - 1]
        for i in range(n - 2, -1, -1):
            suff[i] = nums[i + 1] * suff[i + 1]
        for i in range(n):
            res[i] = pref[i] * suff[i] 
        return res
    

# Prefix & Suffix (Optimal)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * (len(nums))

        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]
        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]
        return res